Source/CombatExtended/CombatExtended/Alerts/Alert_EmptyTurrets.cs

```diff
@@ -0,0 +1,59 @@
+﻿using RimWorld;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using Verse;
+
+namespace CombatExtended.CombatExtended.Alerts
+{
+    class Alert_EmptyTurrets : Alert
+    {
+        private List<Thing> EmptyTurrets
+        {
+            get
+            {
+                List<Thing> emptyTurrets = new List<Thing>();
+                if (Controller.settings.EnableAmmoSystem)
+                {
+                    foreach (Map map in Find.Maps)
+                    {
+                        if (map.IsPlayerHome)
+                        {
+                            foreach (Building_TurretGunCE turretBuilding in map.listerBuildings.AllBuildingsColonistOfClass<Building_TurretGunCE>())
```

Comment:
Alerts need to be *very* efficient, as they are called every frame. You are allocating an object and having two nested foreach loops here. This will almost certainly cause lags.

Reply:
Consider using LINQ instead.

Reply:
Alerts aren't bound to the tick manager, they only execute around twice per second. I'm all for extra performance though, I'll look into LINQ.

Reply:
They aren't called every tick, they're called every *frame*, that's worse.

Reply:
Where are you getting 'every frame' from? That's not how AlertsReadout is set up at all. Pretty sure vanilla's thirty-or-so alerts would slow the game down to a crawl by themselves if they recalculated every frame.